### PR TITLE
Improve YOLO parsing errors

### PR DESCRIPTION
**Pull request by zhiltsov-max**
### Summary

Depends on #686 

- Added more informative YOLO parsing errors
- Added more type annotations in extractor
- Allowed leading spaces in the `.data` file
- Added type docs for `util.image.ImageMeta`

### Checklist

- [ ] I submit my changes into the `develop` branch
- [ ] I have added description of my changes into [CHANGELOG](https://github.com/openvinotoolkit/datumaro/blob/develop/CHANGELOG.md)
- [ ] I have updated the [documentation](
  https://github.com/openvinotoolkit/datumaro/tree/develop/docs) accordingly
- [ ] I have added tests to cover my changes
- [ ] I have [linked related issues](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

### License

- [ ] I submit _my code changes_ under the same [MIT License](
  https://github.com/openvinotoolkit/datumaro/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2021 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

File: datumaro/plugins/yolo_format/extractor.py [hunk 1]
```diff
@@ -2,45 +2,61 @@
...
             self._parent = parent
-            self.items = OrderedDict()
+            self.items: Dict[str, Union[str, DatasetItem]] = OrderedDict()
 
         def __iter__(self):
             for item_id in self.items:
-                yield self._parent._get(item_id, self._name)
+                item = self._parent._get(item_id, self._name)
+                if item is not None:
+                    yield item
```
Comment (IRDonch) [inline datumaro/plugins/yolo_format/extractor.py:42]: This enables questionable semantics: if any items are skipped in this way, then the number of items yielded will be different from the value returned by `len(subset)`.

File: tests/test_yolo_format.py [hunk 2]
```diff
@@ -284,7 +271,7 @@ def test_can_save_and_load_without_path_prefix(self):
                     subset="valid",
                     media=Image(data=np.ones((8, 8, 3))),
                     annotations=[
-                        Bbox(0, 1, 5, 2, label=2),
+                        Bbox(0, 1, 5, 2, label=1),
```
Comment (IRDonch) [inline tests/test_yolo_format.py:274]: What's the reason for this change?

Comment (zhiltsov-max) [reply tests/test_yolo_format.py:274]: The annotation is invalid, there are only 2 labels in the dataset. Now it is checked, and the error is raised.

File: datumaro/plugins/yolo_format/extractor.py (hunk 1 above)
Comment (zhiltsov-max) [reply datumaro/plugins/yolo_format/extractor.py:42]: I agree about this, but it is not clear how to resolve the problem, and it has very limited ways to impact the situation, because it is only used until the imported dataset is cached. During caching/iteration, the actual count is computed. Actually, in VOC we have the same issue. Maybe, we could even consider it normal, because unless we loaded the data, it is the best we can report.

Comment (IRDonch) [reply datumaro/plugins/yolo_format/extractor.py:42]: IMO, the proper solution would involve removing `__len__` from `IExtractor`.

**Review by IRDonch — APPROVED** [review]
